keps/sig-node/kubelet-resource-metrics-endpoint.md

```diff
@@ -0,0 +1,210 @@
+
+#### Methods
+
+The setup has 10 nodes, and up to 40,000 containers (running pause).  Nodes have 2 CPU core, and 7.5Gb memory.  The same cluster was used for all benchmarks for consistency, with a different Metrics Server running.  The values below are the maximum values reported during a 10 minute period.
+
+This experiment "fakes" large numbers of containers by having the kubelet return 100 container metrics for each actual container run on the node.
+
+#### Results
+
+Both gRPC and the optimized prometheus were able to scale to 40k containers.  The gRPC implementation was more efficient by a factor of approx. 3.  
+
+<img src="https://user-images.githubusercontent.com/3262098/51704923-173bd800-1fcf-11e9-910d-3fd6606550f3.png" width="600" height="375">
+
+<img src="https://user-images.githubusercontent.com/3262098/51704880-02f7db00-1fcf-11e9-8034-c64f971a2204.png" width="600" height="375">
+
+## Alternatives Considered
+
+### gRPC API
+
+As demonstrated in the benchmarks above, the proto-based gRPC endpoint is the most efficient in terms of CPU and Memory usage.  Such an endpoint could potentially be improved by using streaming, rather than scraping to be even more efficient at high rates of collection.
+
+However, given the prevalence of the Prometheus format within the kubernetes community, gRPC is not as compatible with common monitoring pipelines.  The endpoint would _only_ be useful for supplying metrics for the Metrics Server, or monitoring components that integrate directly with it.
+
+When using caching in the Metrics Server, the prometheus text format performs _well enough_ for us to prefer prometheus over gRPC given the prevalence of prometheus in the community.  When the OpenMetrics format becomes stable, we can get even closer to the performance of gRPC by using the proto-based format.
+
+```
+// Usage is a set of resources consumed
+message Usage {
+  int64 time = 1;
+  uint64 cpu_usage_core_nanoseconds_total = 2;
+  uint64 memory_working_set_bytes = 3;
+}
+// ContainerUsage is the resource usage for a single container
+message ContainerUsage {
+  string name = 1;
+  Usage usage = 2;
+}
+// PodUsage is the resource usage for a pod
+message PodUsage {
+  string name = 1;
+  string namespace = 2;
+  repeated ContainerUsage containers = 3;
+}
+// MetricsResponse is sent by plugin to kubelet in response to MetricsRequest RPC
+message MetricsResponse {
+  Usage node = 1;
+  repeated PodUsage pods = 2;
+}
+// MetricsRequest is the empty request message for Kubelet
+message MetricsRequest {}
+// ResourceMetrics is the service advertised by the kubelet for usage metrics.
+service ResourceMetrics {
+  rpc Get(MetricsRequest) returns (MetricsResponse) {}
+}
+```
+
+## Graduation Criteria
+
+Alpha:
+
+- [] Implement and test the kubelet resource metrics endpoint as described above
+- [] Modify the metrics server to consume the kubelet resource metrics endpoint
```

Comment:
I can't find whether it is alpha/beta/GA, but I believe it is beta since it is enabled by default in kubernetes clusters.

Reply:
What do you mean by "how will the endpoint be configured"?  Are you asking about something about the metrics server?

Reply:
Yeah. I imagine the metrics server will need to support both summary metric and resource metrics ingestion for a while. I'm assuming this will be some sort of flag or configuration parameter. For heterogeneous clusters with older nodes, I suppose the metrics server will just continue to use the summary endpoint until all nodes support the required version of resource metrics. Just thinking out loud, not sure you need to add this.

Reply:
The current plan is just to switch over once all nodes are expected to have the new endpoint.  I know we only support a 2 version skew between nodes and master version, but i'm not sure how lenient we want to be on enabling a larger skew by providing config to revert to the older API.  It would just be easier for them to run an older metrics server than to add extra configuration...